File: frontend/task-manager-app/src/services/userService.ts
import api from './api';

interface User {
  id: string;
  username: string;
  email: string;
  role: number;
  roleName: string;
}

interface PaginatedResponse<T> {
  items: T[];
  pageNumber: number;
  pageSize: number;
  totalCount: number;
  totalPages: number;
  hasPreviousPage: boolean;
  hasNextPage: boolean;
}

interface UpdateUserRoleRequest {
  role: number;
}

export const userService = {
  async getUsers(pageNumber = 1, pageSize = 10): Promise<PaginatedResponse<User>> {
    const response = await api.get<PaginatedResponse<User>>('/users', {
      params: { pageNumber, pageSize }
    });
    return response.data;
  },

  async getUserById(id: string): Promise<User> {
    const response = await api.get<User>(`/users/${id}`);
    return response.data;
  },

  async updateUserRole(id: string, role: number): Promise<User> {
    const response = await api.put<User>(`/users/${id}/role`, { role } as UpdateUserRoleRequest);
    return response.data;
  },

  async deleteUser(id: string): Promise<void> {
    await api.delete(`/users/${id}`);
  }
};

File: backend/TaskManager.API/Services/Implementations/UserService.cs
using Microsoft.EntityFrameworkCore;
using TaskManager.API.Data;
using TaskManager.API.Exceptions;
using TaskManager.API.Models.DTOs;
using TaskManager.API.Models.Enums;
using TaskManager.API.Services.Interfaces;

namespace TaskManager.API.Services.Implementations
{
    public class UserService : IUserService
    {
        private readonly ApplicationDbContext _context;

        public UserService(ApplicationDbContext context)
        {
            _context = context;
        }

        public async Task<PaginatedResponse<UserDto>> GetUsersAsync(int pageNumber, int pageSize)
        {
            var totalCount = await _context.Users.CountAsync();

            var users = await _context.Users
                .OrderBy(u => u.Username)
                .Skip((pageNumber - 1) * pageSize)
                .Take(pageSize)
                .Select(u => new UserDto
                {
                    Id = u.Id,
                    Username = u.Username,
                    Email = u.Email,
                    Role = u.Role
                })
                .ToListAsync();

            return new PaginatedResponse<UserDto>
            {
                Items = users,
                PageNumber = pageNumber,
                PageSize = pageSize,
                TotalCount = totalCount
            };
        }

        public async Task<UserDto> GetUserByIdAsync(Guid id)
        {
            var user = await _context.Users.FindAsync(id)
                ?? throw new NotFoundException("User", id);

            return new UserDto
            {
                Id = user.Id,
                Username = user.Username,
                Email = user.Email,
                Role = user.Role
            };
        }

        public async Task<UserDto> UpdateUserRoleAsync(Guid id, UserRole newRole)
        {
            var user = await _context.Users.FindAsync(id)
                ?? throw new NotFoundException("User", id);

            user.Role = newRole;
            user.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();

            return new UserDto
            {
                Id = user.Id,
                Username = user.Username,
                Email = user.Email,
                Role = user.Role
            };
        }

        public async Task DeleteUserAsync(Guid id)
        {
            var user = await _context.Users.FindAsync(id)
                ?? throw new NotFoundException("User", id);

            // Prevent deletion if user has tasks
            var hasCreatedTasks = await _context.Tasks.AnyAsync(t => t.CreatedById == id);
            if (hasCreatedTasks)
                throw new BadRequestException("Cannot delete user with existing tasks. Reassign or delete tasks first.");

            _context.Users.Remove(user);
            await _context.SaveChangesAsync();
        }
    }
}